python/ffsim/states/slater.py

```diff
@@ -0,0 +1,296 @@
+            length=len(orbs_b),
+        ), convert_bitstring_type(
+            sampled_configuration_a[:, orbs_a],
+            BitstringType.BIT_ARRAY,
+            bitstring_type,
+            length=len(orbs_b),
+        )
+
+
+def _generate_conditionals_unormalized(
+    rdm: np.ndarray, pos_array: np.ndarray, empty_orbitals: np.ndarray, marginal: float
+) -> tuple[np.ndarray, np.ndarray]:
+    """Generates the conditional and marginal probabilities for adding a particle
+    to the available empty orbitals.
+
+    This is a step of the autoregressive sampling, and uses Bayes's rule.
+
+    Args:
+        rdm: A Numpy array with the one-body reduced density matrix.
+        pos_array: A numpy array with the positions of the particles.
+        empty_orbitals: A numpy array with the empty orbitals that a new particle
+            may occupy. The sorted union of ``pos_array`` and ``empty_orbitals``
+            must be equal to ``numpy.arange(num_orbitals)``.
+        marginal: The marginal probability associated to having the particles
+            in the positions determined by ``pos_array``.
+    Returns:
+        A tuple of Numpy arrays. The first is the unormalized conditionals for
+        adding a particle to any of the empty orbitals specified in the input.
+        The second is the marginal corresponding to having the particles in the
+        position array and one extra in all possible empty orbitals. Both arrays
+        follow the same order as ``empty_orbitals``.
+
+    """
+
+    conditionals = np.zeros(len(empty_orbitals), dtype=complex)
+    marginals = np.zeros(len(empty_orbitals), dtype=complex)
+
+    for i, orbital in enumerate(empty_orbitals):
+        new_pos_array = np.append(pos_array, [orbital])
+        rest_rdm = rdm[new_pos_array, :]
+        rest_rdm = rest_rdm[:, new_pos_array]
+        marginals[i] = np.linalg.det(rest_rdm)
+        conditionals[i] = marginals[i] / marginal
+
+    return conditionals, marginals
+
+
+def _autoregressive_slater(
+    rdm: np.ndarray,
+    norb: int,
+    nelec: int,
+    seed: np.random.Generator | int | None = None,
+) -> np.ndarray:
+    """Autoregressively sample positions of particles for a Slater-determinant wave
+    function using a determinantal point process.
+
+    Args:
+        rdm: A numpy array with the one-body reduced density matrix.
+        norb: Number of orbitals.
+        nelec: Number of electrons.
+        seed: Either a Numpy random generator, an integer seed for the random number
+            generator or ``None``.
+    Returns:
+        A numpy array with the position of the sampled electrons.
+    """
+    rng = np.random.default_rng(seed)
+
+    position = [rng.choice(norb, p=np.real(np.diag(rdm)) / nelec)]
+    marginal = [np.real(np.diag(rdm) / nelec)[position[0]]]
+
+    for k in range(nelec - 1):
+        pos_array = np.array(position, dtype=int)
+
+        empty_orbitals = np.setdiff1d(np.arange(norb), pos_array, assume_unique=True)
```

Comment:
As I mention above, I think we should simply use Python lists for `pos_array` and `empty_orbitals` instead of Numpy arrays.